File: core/tests/fixtures/mothers/base/base_mother.py
from abc import ABC
from typing import Any, Callable, Optional

from dateutil.parser import parse

from .mother_mapping import MotherMapping


class BaseMother(ABC):
    @classmethod
    def _merge(cls, base: dict, overrides: dict) -> dict:
        """Merges the base dictionary with the overrides."""
        values = base.copy()

        for key, value in values.items():
            if key in overrides and overrides[key] is not None:
                values[key] = overrides[key]
                continue

            (value, mapping, builder) = cls._get_mapping(value)
            values[key] = cls._apply_mapping(value, mapping, builder)

        return values

    @staticmethod
    def _get_mapping(item: Any) -> tuple[Any, MotherMapping, Optional[Callable]]:
        """Returns the value, mapping, and builder for the given value."""
        if not isinstance(item, list):
            if callable(item):
                return None, MotherMapping.MOTHER, item
            return item, MotherMapping.NONE, None

        value = None
        mapping = MotherMapping.NONE
        builder = None

        for element in item:
            if isinstance(element, MotherMapping):
                mapping = element
            elif callable(element):
                builder = element
            else:
                value = element

        if builder is not None:
            mapping = MotherMapping.MOTHER
            value = None

        return value, mapping, builder

    @staticmethod
    def _apply_mapping(value: Any, mapping: MotherMapping, builder: Optional[Callable]) -> Any:
        """Applies the mapping to the value using the builder."""
        if mapping == MotherMapping.REQUIRED and value is None:
            raise ValueError("Value is required but not provided.")

        if mapping == MotherMapping.MOTHER and builder is not None and value is None:
            return builder()

        if mapping == MotherMapping.DATE and isinstance(value, str):
            return parse(value)

        return value

File: editor/tests/fixtures/mothers/base/base-mother.ts
import { isMotherMapping, MotherMapping } from './mother-mapping';
import { MotherBuild } from './mother-build';

type ItemValue =
  | string
  | number
  | boolean
  | Date
  | MotherMapping
  | MotherBuild
  | any[]
  | null;
export type Item = ItemValue | ItemValue[];

export abstract class BaseMother {
  protected static merge<T = object, U = T>(
    values: T,
    overrides?: Partial<U>,
  ): U {
    const fields: U = {} as U;

    for (const key of Object.keys(values as object)) {
      // @ts-ignore
      if (overrides && key in overrides && overrides[key] !== undefined) {
        // @ts-ignore
        fields[key] = overrides[key] as U[Extract<keyof U, string>];
        continue;
      }

      // @ts-ignore
      const item = values[key] as Item;
      const [value, mapping, builder] = BaseMother.getMapping(item);
      // @ts-ignore
      fields[key] = BaseMother.applyMapping(value, mapping, builder);
    }

    return fields;
  }

  protected static getMapping(
    item: Item,
  ): [ItemValue, MotherMapping, MotherBuild | null] {
    if (!Array.isArray(item)) {
      if (typeof item === 'function') {
        return [null, MotherMapping.MOTHER, new MotherBuild(item)];
      }
      if (typeof item === 'string' && isMotherMapping(item)) {
        return [null, item, null];
      }

      return [item, MotherMapping.NONE, null];
    }

    let value: ItemValue = null;
    let map = MotherMapping.NONE;
    let builder: MotherBuild | null = null;

    for (const element of item) {
      if (typeof element === 'string' && isMotherMapping(element)) {
        map = element;
      } else if (typeof element === 'function') {
        map = MotherMapping.MOTHER;
        builder = new MotherBuild(element);
      } else {
        value = element;
      }
    }

    if (null !== builder) {
      value = null;
    }

    return [value, map, builder];
  }

  protected static applyMapping(
    value: ItemValue,
    mapping: MotherMapping,
    builder: MotherBuild | null,
  ): ItemValue {
    if (mapping === MotherMapping.REQUIRED && !value) {
      throw new Error('Value is required');
    }

    if (
      mapping === MotherMapping.MOTHER &&
      null !== builder &&
      null === value
    ) {
      return builder.build() as ItemValue;
    }

    if (mapping === MotherMapping.ARRAY && Array.isArray(value)) {
      return value.map((item) => {
        if (typeof item === 'function') {
          return item();
        }
        return item;
      }) as ItemValue;
    }

    if (mapping === MotherMapping.DATE) {
      if (!value) {
        return new Date();
      }

      return new Date(value as string);
    }

    if (mapping === MotherMapping.UUID) {
      return crypto.randomUUID();
    }

    return value;
  }
}
